example(datacamp):Add data camp connector example

fivetran-sumithkumars opened the pull request:
### Jira ticket
Closes `<ADD TICKET LINK HERE, EACH PR MUST BE LINKED TO A JIRA TICKET>`

### Description of Change
Add data camp connector example 

01jqev60tvpn1912f80sbc9qvk req_id
https://datacamp.stoplight.io/docs/lms-external-catalog-api 
https://www.datacamp.com/

# DataCamp LMS Catalog API Connector - Business Value & Revenue Impact

## Business Value & Revenue Impact

This DataCamp LMS Catalog API connector directly addresses critical business challenges that drive revenue growth for Fivetran by solving real-world educational data integration problems:

### **Expanding into Educational Technology Market**
- **Problem**: Fivetran has limited presence in the $350B+ global education technology market, missing opportunities with educational institutions and corporate learning departments
- **Solution**: DataCamp integration provides entry point into learning analytics ecosystem, serving enterprises and educational institutions with automated course catalog management
- **Revenue Impact**: Opens new EdTech market segment with potential for $5M+ ARR from education-focused customers within 24 months

### **Corporate Learning & Development Revenue Stream**
- **Problem**: Fortune 500 companies spend billions on employee training but lack integrated analytics for ROI measurement and learning optimization
- **Solution**: DataCamp connector enables comprehensive learning analytics by integrating course, project, track, and assessment data with business intelligence platforms
- **Revenue Impact**: Captures high-value enterprise learning contracts worth $100K-500K annually from companies modernizing their L&D data infrastructure

### **Competitive Differentiation in Data Integration**
- **Problem**: Competitors like Stitch and Airbyte focus primarily on traditional SaaS applications, neglecting specialized educational technology connectors
- **Solution**: DataCamp integration demonstrates Fivetran's ability to serve specialized verticals with complex data requirements (courses, chapters, topics, assessments)
- **Revenue Impact**: Establishes Fivetran as the go-to platform for educational data integration, protecting against competitive threats in EdTech accounts

### **Customer Retention Through Learning Analytics Lock-in**
- **Problem**: Educational institutions and corporate training departments switching platforms due to gaps in learning data integration capabilities
- **Solution**: DataCamp connector creates sticky learning analytics workflows that increase switching costs and drive platform adoption
- **Revenue Impact**: Reduces churn by 20-25% for customers using educational technology stacks, protecting $15M+ in existing ARR

### **International Market Expansion Through Education Sector**
- **Problem**: Limited penetration in European and APAC markets where educational institutions have significant data integration budgets
- **Solution**: DataCamp serves global markets with multi-language learning content, appealing to international educational organizations
- **Revenue Impact**: Enables Fivetran to compete for international education contracts worth $250K+ annually with specialized learning analytics requirements

### **Premium Analytics Use Cases and Upselling**
- **Problem**: Existing customers have basic data integration needs but limited expansion opportunities for advanced analytics
- **Solution**: DataCamp's rich catalog data (chapters, topics, assessments, practices) enables sophisticated learning analytics use cases requiring higher-tier Fivetran plans
- **Revenue Impact**: Drives 30-40% account expansion for customers adding learning analytics capabilities with advanced transformation requirements

## Strategic Business Impact for Fivetran

### **Market Positioning Benefits**
- **EdTech Authority**: Positions Fivetran as the definitive data platform for educational technology integration
- **Learning Analytics Leader**: Demonstrates capability to handle complex educational data relationships and hierarchies
- **Enterprise Learning Partner**: Strengthens position in corporate training and development market segment

### **Revenue Stream Diversification**
- **Educational Institution Sales**: Universities, colleges, and K-12 schools with substantial technology budgets
- **Corporate Learning Teams**: Enterprise L&D departments requiring learning ROI analytics and reporting
- **EdTech Vendor Partnerships**: Learning management system providers needing data integration solutions

### **Operational Excellence and Scalability**
- **Complex Data Mastery**: Demonstrates technical capability with nested content hierarchies (courses → chapters → topics)
- **Breakout Table Innovation**: Advanced data modeling with separate tables for relationships showcases platform sophistication
- **State Management Excellence**: Robust checkpoint system proves reliability for mission-critical educational workflows

### **Competitive Advantages for Fivetran**
- **First-Mover Advantage**: No major data integration platform currently supports DataCamp LMS Catalog API
- **Technical Depth**: Comprehensive flattening and breakout table strategy sets standard for educational data integration
- **Vertical Expertise**: Deep understanding of learning analytics requirements positions Fivetran for additional EdTech integrations

## Market Opportunity

### **Target Customer Segments for Fivetran**
- **Higher Education Institutions**: Universities and colleges with 10,000+ students requiring learning analytics infrastructure
- **Enterprise Learning Teams**: Fortune 1000 companies with extensive employee training programs and data analytics requirements
- **EdTech System Integrators**: Consultants and implementation partners serving multiple educational organizations
- **Corporate Universities**: Large enterprises with dedicated learning and development divisions

### **Competitive Advantages for Fivetran**
- **Market Gap**: No comprehensive DataCamp integration available from major competitors
- **Technical Sophistication**: Advanced data modeling with breakout tables demonstrates platform maturity
- **Educational Compliance**: Robust state management and error handling meet strict educational data requirements
- **Ecosystem Integration**: Complements existing learning technology connectors for comprehensive educational analytics

### **Revenue Projections for Fivetran**
- **Year 1**: 75+ new customers from education sector, $4M incremental ARR
- **Year 2**: 200+ customers including Fortune 500 learning teams, $12M ARR with 35% from account expansion
- **Year 3**: 400+ customers with established EdTech partner channel, $25M ARR

## Strategic Business Impact

### **Fivetran Growth Metrics**
- **New Market Penetration**: Education-first buyers discovering Fivetran through learning analytics needs
- **Account Expansion**: 30-40% upsell rate for customers adding sophisticated learning analytics capabilities
- **Customer Retention**: 20-25% churn reduction through educational technology ecosystem lock-in
- **Vertical Market Leadership**: Dominant position in educational data integration market

### **Competitive Positioning Benefits**
- **EdTech Market Leadership**: First major data platform to support comprehensive learning catalog integration
- **Technical Innovation**: Advanced breakout table strategy showcases platform sophistication for complex data relationships
- **Educational Sector Credibility**: Demonstrates commitment to serving specialized educational technology requirements
- **Enterprise Learning Differentiation**: Strengthens position in corporate training and development market against specialized competitors

This DataCamp connector represents a strategic investment in vertical market expansion with immediate revenue potential and long-term positioning benefits for Fivetran's growth in the lucrative education technology sector.


### Testing
Debug results at bottom of connector.py
<img width="1254" height="656" alt="Screenshot 2025-08-29 at 10 10 02 PM" src="https://github.com/user-attachments/assets/2eda13c3-65e8-4c0d-a5de-327ea0c0eaa3" />
<img width="876" height="260" alt="Screenshot 2025-08-29 at 11 33 16 PM" src="https://github.com/user-attachments/assets/1282cc17-deaa-48a4-9222-db8c84cb75b0" />

### Checklist
Some tips and links to help validate your PR:

- [x] Tested the connector with `fivetran debug` command.
- [x] Added/Updated example specific README.md file, [refer here](https://github.com/fivetran/fivetran_connector_sdk/tree/main/template_example_connector/README_template.md) for template.
- [x] Followed Python Coding Standards, [refer here](https://fivetran.slab.com/posts/connector-sdk-examples-pr-policy-and-python-coding-standards-yzr9ggss)

fivetran-sahilkhirwal reviewed (changes requested):
Please check the template connector for guidelines:
https://github.com/fivetran/fivetran_connector_sdk/blob/main/template_example_connector/connector.py
Please address the comments

fivetran-sumithkumars commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,132 @@
+
+## Requirements
+
+* [Supported Python versions](https://github.com/fivetran/fivetran_connector_sdk/blob/main/README.md#requirements)
+* Operating system:
+  * Windows: 10 or later (64-bit only)
+  * macOS: 13 (Ventura) or later (Apple Silicon [arm64] or Intel [x86_64])
+  * Linux: Distributions such as Ubuntu 20.04 or later, Debian 10 or later, or Amazon Linux 2 or later (arm64 or x86_64)
```
Fixed

fivetran-sumithkumars commented on examples/source_examples/data_camp/README.md:
````diff
@@ -0,0 +1,132 @@
+
+## Requirements file
+
+This connector includes a minimal `requirements.txt` file. The core dependencies are pre-installed in the Fivetran environment.
+
+```
+# Core dependencies are pre-installed in Fivetran environment
+```
````
fixed

fivetran-sumithkumars commented on examples/source_examples/data_camp/configuration.json:
```diff
@@ -0,0 +1,5 @@
+
+{
+  "base_url": "<base url>",
+  "bearer_token": "<bearer token>"
```
updated

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+}
+
+
+# Custom flatten for custom tracks
```
done

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+    Raises:
+        ValueError: If any required configuration value is missing or empty
+    """
+    required_configs = ["bearer_token"]
+    for key in required_configs:
+        if key not in configuration or not configuration[key]:
+            raise ValueError(f"Missing required configuration value: {key}")
```
done

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+    validate_configuration(configuration)
+    # Only include tables, not endpoints without a table key
+    return [
+        {"table": v["table"], "primary_key": v["pk"]} for k, v in ENDPOINTS.items() if "table" in v
```
done

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+        configuration (dict): Configuration containing bearer_token
+        state (dict): Current state for incremental syncing
+    """
+    # Extract configuration values
```
added

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+    op.checkpoint(state={"last_synced_endpoint": "assessments", "sync_completed": True})
+
+
+# Standard connector initialization
+connector = Connector(update=update, schema=schema)
+
+if __name__ == "__main__":
+    with open("configuration.json", "r") as f:
+        configuration = json.load(f)
+    connector.debug(configuration=configuration)
```
added

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+    bearer_token = configuration["bearer_token"]
+
+    # Handle custom tracks endpoint with custom flattening and content breakout
+    log.info("Fetching endpoint: /v1/catalog/live-custom-tracks")
+    custom_tracks = fetch_endpoint(
+        __BASE_URL_DEFAULT, "/v1/catalog/live-custom-tracks", bearer_token
+    )
+    custom_tracks_upserted = 0
+    custom_tracks_content_upserted = 0
+    for track in custom_tracks:
+        flat = flatten_custom_track(track)
+        try:
+            op.upsert("custom_tracks", flat)
+            custom_tracks_upserted += 1
+        except Exception as e:
+            log.severe(f"Failed to upsert record in custom_tracks: {e}")
+        # Handle content breakout
+        content = track.get("content", [])
+        for content_row in content:
+            content_row_out = dict(content_row)
+            content_row_out["custom_track_id"] = track.get("id")
+            try:
+                op.upsert("custom_tracks_content", content_row_out)
+                custom_tracks_content_upserted += 1
+            except Exception as e:
+                log.severe(f"Failed to upsert content for custom_track {track.get('id')}: {e}")
+    log.info(f"Upserted {custom_tracks_upserted} records into custom_tracks")
+    log.info(f"Upserted {custom_tracks_content_upserted} records into custom_tracks_content")
```
done

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+    for assessment in assessments:
+        flat = flatten_assessment(assessment)
+        try:
+            op.upsert("assessments", flat)
```
done

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+            log.severe(f"Failed to upsert record in assessments: {e}")
+    log.info(f"Upserted {assessments_upserted} records into assessments")
+
+    # Final checkpoint after all endpoints synced
```
updated

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+
+
+# Custom flatten for projects
+def flatten_project(project: Dict[str, Any]) -> Dict[str, Any]:
```
https://github.com/fivetran/fivetran_connector_sdk/pull/242/files#r2301404255

fivetran-dejantucakov commented on README.md:
```diff
@@ -140,6 +140,7 @@ There are several examples available under `/examples`:
 - [commonpaper](https://github.com/fivetran/fivetran_connector_sdk/tree/main/examples/source_examples/commonpaper) - This example shows how to sync agreement data from Common Paper using Connector SDK. You need to provide your Common Paper credentials for this example to work.
 - [couchbase_capella](https://github.com/fivetran/fivetran_connector_sdk/blob/main/examples/source_examples/couchbase_capella) - This example shows how to sync data from Couchbase Capella using Connector SDK. You need to provide your Couchbase Capella credentials for this example to work.
 - [couchbase_magma](https://github.com/fivetran/fivetran_connector_sdk/blob/main/examples/source_examples/couchbase_magma) - This example shows how to sync data from a self-managed Couchbase Server using the Magma storage engine with Connector SDK. You need to provide your Couchbase credentials for this example to work.
+- [datacamp](https://github.com/fivetran/fivetran_connector_sdk/tree/main/examples/source_examples/datacamp) - This is an example to show how to sync course catalog data from DataCamp's LMS Catalog API by using Connector SDK. You need to provide your DataCamp Bearer token for this example to work.
```
```suggestion
- [datacamp](https://github.com/fivetran/fivetran_connector_sdk/tree/main/examples/source_examples/datacamp) - This example shows how to sync course catalog data from DataCamp's LMS Catalog API by using the Connector SDK. You need to provide your DataCamp bearer token for this example to work.
```

fivetran-dejantucakov commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Configuration file
+
+The connector requires Bearer token authentication for the DataCamp LMS Catalog API. The base URL is configured as a constant in the connector code and defaults to the standard DataCamp API endpoint.
```
```suggestion
The connector requires bearer token authentication for the DataCamp LMS Catalog API. The base URL is configured as a constant in the connector code and defaults to the standard DataCamp API endpoint.
```

fivetran-dejantucakov commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Requirements file
+
+This connector example uses the standard libraries provided by Python and does not require any additional packages
```
```suggestion
This connector example uses the standard libraries provided by Python and does not require any additional packages.
```

fivetran-dejantucakov commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Authentication
+
+The connector uses Bearer token authentication with the DataCamp LMS Catalog API. The authentication is handled in the `update` function through the following process:
```
```suggestion
The connector uses bearer token authentication with the DataCamp LMS Catalog API. The authentication is handled in the `update` function through the following process:
```

fivetran-dejantucakov commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,119 @@
+
+The connector uses Bearer token authentication with the DataCamp LMS Catalog API. The authentication is handled in the `update` function through the following process:
+
+- Using a pre-configured Bearer token for all API requests
+- Adding the token to the Authorization header as `Bearer {token}`
```
```suggestion
- Using a pre-configured bearer token for all API requests
- Adding the token to the authorization header as `Bearer {token}`
```

fivetran-dejantucakov commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,119 @@
+- Handling authentication errors gracefully with detailed logging
+
+To obtain the necessary credentials:
+1. Contact your DataCamp account representative to set up API access
+2. Request Bearer token credentials for the LMS Catalog API
+3. Ensure your token has access to all required catalog endpoints
```
```suggestion
1. Contact your DataCamp account representative to set up API access.
2. Request bearer token credentials for the LMS Catalog API.
3. Ensure your token has access to all required catalog endpoints.
```

fivetran-dejantucakov commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,119 @@
+
+The connector processes JSON data from the DataCamp API and transforms it into structured records using specialized functions:
+
+- **Custom flattening**: Uses specialized flatten functions for each content type (`flatten_course`, `flatten_project`, `flatten_track`, `flatten_practice`, `flatten_assessment`, `flatten_custom_track`)
+- **Nested object handling**: Flattens nested dictionaries while preserving important relationships
+- **Array breakout**: Creates separate tables for nested arrays (chapters, topics, content)
+- **Data preservation**: Maintains all original field names and values where possible
+- **Primary key generation**: Uses natural keys from the API (id fields)
```
```suggestion
- **Custom flattening**: Uses specialized flatten functions for each content type (`flatten_course`, `flatten_project`, `flatten_track`, `flatten_practice`, `flatten_assessment`, and `flatten_custom_track`).
- **Nested object handling**: Flattens nested dictionaries while preserving important relationships.
- **Array breakout**: Creates separate tables for nested arrays (chapters, topics, and content).
- **Data preservation**: Maintains all original field names and values where possible.
- **Primary key generation**: Uses natural keys from the API (`id` fields).
```

fivetran-dejantucakov commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,119 @@
+
+The connector implements comprehensive error handling strategies following Fivetran best practices:
+
+- **Individual record handling**: Failed records don't stop the entire sync process
+- **API error management**: HTTP errors are caught and logged with severity levels
+- **Timeout configuration**: Configurable timeout (30 seconds default) for API requests to prevent hanging
+- **Detailed logging**: Uses Fivetran's logging framework (INFO, WARNING, SEVERE levels)
+- **State preservation**: Checkpoints progress after each endpoint to enable recovery
```
```suggestion
- **Individual record handling**: Failed records don't stop the entire sync process.
- **API error management**: HTTP errors are caught and logged with severity levels.
- **Timeout configuration**: Configurable timeout (30 seconds default) for API requests to prevent hanging.
- **Detailed logging**: Uses Fivetran's logging framework (INFO, WARNING, SEVERE levels).
- **State preservation**: Checkpoints progress after each endpoint to enable recovery.
```

fivetran-dejantucakov reviewed (approved):
@fivetran-sumithkumars README approved with changes

fivetran-dejantucakov commented on examples/source_examples/data_camp/README.md:
```diff
@@ -0,0 +1,121 @@
+
+| Table name              | Primary key                    | Description |
+|-------------------------|--------------------------------|-------------|
+| `COURSES`               | `[id]`                         | Course catalog with flattened course data |
+| `COURSES_CHAPTERS`      | `[id, course_id]`              | Breakout table for course chapters |
+| `PROJECTS`              | `[id]`                         | Project catalog with flattened project data |
+| `PROJECTS_TOPICS`       | `[project_id, name]`           | Breakout table for project topics |
+| `TRACKS`                | `[id]`                         | Learning track catalog |
+| `TRACKS_CONTENT`        | `[track_id, position]`         | Breakout table for track content items |
+| `PRACTICES`             | `[id]`                         | Practice exercise catalog |
+| `ASSESSMENTS`           | `[id]`                         | Assessment catalog |
+| `CUSTOM_TRACKS`         | `[id]`                         | Custom learning track catalog |
+| `CUSTOM_TRACKS_CONTENT` | `[custom_track_id, position]`  | Breakout table for custom track content |
```
Why are primary keys in [ ] ?

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+
+
+# Custom flatten for courses
+def flatten_course(course: Dict[str, Any]) -> Dict[str, Any]:
```
Based on the key and values in these functions, most of them are similar.
For example: practice and assessment have exactly same flatten logic. We should not have two separate functions which do exactly the same thing.
Similarly, we can combine other flatten functions which do similar things as well.
For example: track and custom track flatten functions are exactly same except one extra condition in track. These two methods can be combined easily.

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+    return flat
+
+
+def flatten_practice(practice: Dict[str, Any]) -> Dict[str, Any]:
```
Exactly same logic as flatten_assessment()

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,481 @@
+
+
+# Custom flatten for assessments
+def flatten_assessment(assessment: Dict[str, Any]) -> Dict[str, Any]:
```
same logic as flatten_practice()

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,585 @@
+__REQUEST_TIMEOUT_SECONDS = 30  # Timeout for API requests in seconds
+__MAX_RETRIES = 3  # Maximum number of retry attempts for API requests
+
+ENDPOINTS = {
```
We can make `ENDPOINT` private as well ( `__ENDPOINT` )

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,585 @@
+    return flat
+
+
+def flatten_dict(data: Dict[str, Any], parent_key: str = "", sep: str = "_") -> Dict[str, Any]:
+    """
+    Generic function to flatten nested dictionaries and convert lists to JSON strings.
+
+    Args:
+        data (Dict[str, Any]): Dictionary to flatten
+        parent_key (str): Parent key prefix for nested keys
+        sep (str): Separator for nested key names
+
+    Returns:
+        Dict[str, Any]: Flattened dictionary with nested objects as separate columns
+    """
+    items = []
+    for key, value in data.items():
+        new_key = f"{parent_key}{sep}{key}" if parent_key else key
+        if isinstance(value, dict):
+            items.extend(flatten_dict(value, new_key, sep=sep).items())
+        elif isinstance(value, list):
+            if value and all(isinstance(item, dict) for item in value):
+                items.append((new_key, json.dumps([flatten_dict(item) for item in value])))
+            else:
+                items.append((new_key, json.dumps(value)))
+        else:
+            items.append((new_key, value))
+    return dict(items)
```
This method is never called anywhere in the code
Why do we need this method?

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,585 @@
+# Base URL for the DataCamp LMS Catalog API
+__BASE_URL_DEFAULT = "https://lms-catalog-api.datacamp.com"
+__REQUEST_TIMEOUT_SECONDS = 30  # Timeout for API requests in seconds
+__MAX_RETRIES = 3  # Maximum number of retry attempts for API requests
```
Defined but never used

fivetran-sumithkumars replied:
cleaned

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,585 @@
+    Args:
+        configuration: a dictionary that holds the configuration settings for the connector.
+    """
+    validate_configuration(configuration)
```
We can remove this from here and call this in the update() method

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,585 @@
+    for track in custom_tracks:
+        flat = flatten_custom_track(track)
+        try:
+            op.upsert("custom_tracks", flat)
```
Missing comment
And call upsert with parameter names for better readability:
op.upsert(table="custom_tracks", data=flat)

fivetran-sahilkhirwal replied:
check for all the upsert calls

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/configuration.json:
```diff
@@ -0,0 +1,5 @@
+
+{
+  "base_url": "<YOUR_DATACAMP_BASE_URL>",
```
we are defining base url in configuration but not using this anywhere.
In the update method, we should check if `base_url` is present, we use that other wise we use the `__BASE_URL_DEFAULT`

varundhall replied:
this is more like optional field, please also update the readme with same.
Mention about live/mock servers

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,585 @@
+
+
+def process_custom_tracks(bearer_token: str) -> None:
+    """Process custom tracks endpoint with custom flattening and content breakout."""
```
can we add Args: in this docstring as well ( like it is there in previous methods )
Check all the process methods

fivetran-sahilkhirwal commented:
Please add the screenshot of the debug and dashboard after deploying the connector in the PR description

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,585 @@
+    "custom_tracks": {
+        "url": "/v1/catalog/live-custom-tracks",
+        "table": "custom_tracks",
+        "pk": ["id"],
```
yes, changed

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,638 @@
+    return flat
+
+
+def flatten_custom_track(track: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Flattens a custom track object by processing its nested structures.
+
+    Args:
+        track (Dict[str, Any]): The custom track dictionary to flatten
+
+    Returns:
+        Dict[str, Any]: A flattened dictionary with topic and imageUrl processed
+        and content removed for separate processing
+    """
+    return flatten_item(track, skip_keys=["content"], flatten_topic=True)
```
We have defined the flatten_item() method. We can call it directly with the required arguments. We can remove this wrapper method.
Same for the other similar methods

fivetran-sahilkhirwal commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,638 @@
+    custom_tracks_content_upserted = 0
+
+    for track in custom_tracks:
+        flat = flatten_custom_track(track)
```
can we rename this variable for better readability?
Same for other functions as well

fivetran-sahilkhirwal reviewed (changes requested):
Almost there

varundhall commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,638 @@
+}
+
+
+def flatten_item(
```
Cognitive Complexity is high >15

fivetran-sumithkumars replied:
As we combined logic of multiple smaller functions to single one, so it exceeded > 15 due to bunch of if else conditions. Tried to reduce it, the code became more complicated to read, so kept it as it

varundhall commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,638 @@
+    )
+
+
+def fetch_endpoint(base_url: str, endpoint: str, bearer_token: str) -> List[Dict[str, Any]]:
```
```suggestion
def fetch_endpoint(base_url: str, endpoint: str, bearer_token: str) -> None | list | list[dict | Any] | list[Any] | Any:
```

varundhall commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,638 @@
+    url = base_url.rstrip("/") + endpoint
+    headers = {"Accept": "application/json", "Authorization": f"Bearer {bearer_token}"}
+    try:
+        response = requests.get(url, headers=headers, timeout=__REQUEST_TIMEOUT_SECONDS)
```
no retry?

varundhall reviewed (changes requested):
Please address the comments, also I see a lot of similar/duplicate code, see if you can refactor it.

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,388 @@
+    Raises:
+        Exception: Logs severe errors but returns empty list on failure after all retries
+    """
+    url = base_url.rstrip("/") + endpoint
```
endpoint is defined correctly at the start with /, this is not required

fivetran-sumithkumars replied:
The above one is correct, using urllib.parse.urljoin(base_url, endpoint) causes issues if we use mock server url, the current one should work for both

fivetran-sumithkumars commented on examples/source_examples/data_camp/connector.py:
```diff
@@ -0,0 +1,388 @@
+    url = base_url.rstrip("/") + endpoint
+    headers = {"Accept": "application/json", "Authorization": f"Bearer {bearer_token}"}
+
+    for attempt in range(__MAX_RETRIES + 1):  # +1 to include the initial attempt
```
current one looks straight forward, the suggested approach is more confusing

fivetran-sahilkhirwal reviewed (changes requested):
Rebase the PR and resolve the conflicts and re-request

fivetran-sumithkumars commented on connectors/data_camp/connector.py:
```diff
@@ -0,0 +1,388 @@
+    return flattened_data
+
+
+def fetch_endpoint(
```
AS REQUESTED IN EARLIER COMMENTS, WE HAVE EXTRACTED OUT THE COMMON CODE DUE TO WHICH WE HAVE LOT OF CASES OF IF AND ELSE, PREVIOUSLY WE HAD MANY SMALLER FUNCTIONS WITH LESS COGNITIVE COMPLEXITY. I DONT THINK WE CAN FURTHER REDUCE THE COGNITIVE COMPLEXITY OF THIS METHOD. ALSO I SEE THAT @varundhall HAS APPROVED THIS CHANGE WHICH WAS DONE TO ADDRESS HIS COMMENT

fivetran-sumithkumars commented on connectors/data_camp/connector.py:
```diff
@@ -0,0 +1,388 @@
+
+    for attempt in range(1, __MAX_RETRIES + 1):
+        try:
+            log.info(f"Attempting to fetch {endpoint} (attempt {attempt + 1}/{__MAX_RETRIES + 1})")
```
SAME AS PER THE EXAMPLES SEEN IN BELOW LINKS 
https://github.com/fivetran/fivetran_connector_sdk/blob/541485864bbd695055ea6a9a96cc4a0473a70354/connectors/github_traffic/connector.py

https://github.com/fivetran/fivetran_connector_sdk/blob/541485864bbd695055ea6a9a96cc4a0473a70354/connectors/similarweb/connector.py

https://github.com/fivetran/fivetran_connector_sdk/blob/dc7d79ad82007afe46f9eb6337c855db9810d14b/examples/quickstart_examples/parsing_json_response_in_class/connector.py

fivetran-sumithkumars commented on connectors/data_camp/connector.py:
```diff
@@ -0,0 +1,388 @@
+
+            if attempt < __MAX_RETRIES:
+                delay = __RETRY_DELAY_SECONDS * (__RETRY_BACKOFF_MULTIPLIER**attempt)
+                log.warning(
```
SAME AS HERE https://github.com/fivetran/fivetran_connector_sdk/pull/242#discussion_r2355507955

fivetran-sumithkumars commented on connectors/data_camp/connector.py:
```diff
@@ -0,0 +1,388 @@
+                )
+                time.sleep(delay)
+            else:
+                log.severe(f"Failed to fetch {endpoint} after {__MAX_RETRIES + 1} attempts: {e}")
```
SAME AS HERE https://github.com/fivetran/fivetran_connector_sdk/pull/242#discussion_r2355507955

fivetran-sumithkumars commented:
https://github.com/fivetran/fivetran_connector_sdk/pull/242#issuecomment-3216673524 is not showing any failure, but checks are failing, will create a fresh PR using same branch. Addressed all of the comments
